### PR TITLE
Handle ScalarValue::Dictionary in add_to_row and update_avg_to_row

### LINKED ISSUE #6174: Handle ScalarValue::Dictionary in add_to_row and update_avg_to_row
### Describe the bug

We've encountered errors about Dictionary scalar that sum row updater cannot handle, while running some internal tests on top of DataFusion. Under some conditions, hash aggregation updates accumulators using scalar from input arrays, the scalar values obtained from `try_from_array` could be Dictionary scalar. Currently seems `add_to_row` and `update_avg_to_row` both don't handle such cases.

### To Reproduce

_No response_

### Expected behavior

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by viirya**
# Which issue does this PR close?

Closes #6174.

# Rationale for this change

# What changes are included in this PR?

# Are these changes tested?

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/physical-expr/src/aggregate/sum.rs [hunk 1]
```diff
@@ -546,4 +555,76 @@ mod tests {
...
+        let mut accum = agg.create_row_accumulator(0)?;
+        let scalar_value = ScalarValue::try_from_array(array, 0)?;
+
+        accum.update_scalar(&scalar_value, row_accessor)?;
+        accum.evaluate(row_accessor)
+    }
+
+    #[test]
+    fn sum_dictionary_f64() -> Result<()> {
+        let keys = Int32Array::from(vec![0, 1, 2, 3, 4]);
```
Comment (alamb) [inline datafusion/physical-expr/src/aggregate/sum.rs:573]: I recommend using keys that are different from values, to help improve the coverage.

Also perhaps a repeated value rather that distinct values would help

**Review by alamb — APPROVED** [review]
The code looks good to me. Thank you @viirya 

I had some minor test suggestions but I don't think that would block merging

My biggest feedback is that it would also be great to add an end to end sql test of this (as sql logic test). Sadly, I couldn't find any good existing example in 
https://github.com/apache/arrow-datafusion/tree/main/datafusion/core/tests/sqllogictests/test_files

However here is an example that we perhaps could adapt:

```sql
❯ create table value(x bigint) as values (1), (2), (3), (4), (5);
0 rows in set. Query took 0.006 seconds.
❯ create table value_dict as select arrow_cast(x, 'Dictionary(Int64, Int32)') as x_dict from value;
0 rows in set. Query took 0.003 seconds.
❯ select * from value_dict;
+--------+
| x_dict |
+--------+
| 1      |
| 2      |
| 3      |
| 4      |
| 5      |
+--------+
5 rows in set. Query took 0.002 seconds.
❯ select sum(x_dict) from value_dict;
Error during planning: The function Sum does not support inputs of type Dictionary(Int64, Int32).
❯ select avg(x_dict) from value_dict;
Error during planning: The function Avg does not support inputs of type Dictionary(Int64, Int32).
❯ select min(x_dict) from value_dict;
+------------------------+
| MIN(value_dict.x_dict) |
+------------------------+
| 1                      |
+------------------------+
1 row in set. Query took 0.005 seconds.
❯ select max(x_dict) from value_dict;
+------------------------+
| MAX(value_dict.x_dict) |
+------------------------+
| 5                      |
+------------------------+
1 row in set. Query took 0.004 seconds.
```


cc @yjshen

File: datafusion/physical-expr/src/aggregate/sum.rs (hunk 1 above)
Comment (viirya) [reply datafusion/physical-expr/src/aggregate/sum.rs]: Yea, improved the test coverage.

**Comment by viirya** [comment]
> Error during planning: The function Sum does not support inputs of type Dictionary(Int64, Int32).

Different to the physical expression issue in this PR, this error looks like a separate issue happened in query planing. We don't encounter it as we don't rely on the query planning in DataFusion. Let me take a look of it in a separate PR.

**Comment by alamb** [comment]
> >  My biggest feedback is that it would also be great to add an end to end sql test of this (as sql logic test). Sadly, I couldn't find any good existing example in

> I think we can definitely an end to end sql test of this. But due to the planning issue above, we cannot run into the scalar update path as the query cannot be planning. I will address this planning issue and add end to end sql tests after it.

Makes sense to me -- I find that the end to end sql tests sometimes catch additional bugs that unit / programatic ones do not

**Comment by viirya** [comment]
Thank you @alamb. I will submit a PR for the planning issue later.

**Pull request merged**